### ISSUE TITLE
Rust: sync everything from crates.io

### ISSUE THREAD

**Issue by olafurpg** [sourcegraph/sourcegraph#37921]
This issue is about tracking progress towards syncing everything (~86k crates, 1.3m versions) from crates.io

- [x] Sync 10 crates from olafurpg/crates-testing-index
- [x] Sync 5k crates from olafurpg/crates.io-index
- [x] Sync 87k crates from rust-lang/crates.io-index
- [x] Fix regressions in npm/python package hosts https://github.com/sourcegraph/sourcegraph/pull/38652
- [x] Enable continuous syncing with rust-lang/crates.io-index
- [x] https://github.com/sourcegraph/sourcegraph/pull/38811
- [x] https://github.com/sourcegraph/sourcegraph/issues/37690
- [ ] Support lazy syncing of new crates that have no previous releases

**Comment by olafurpg** [comment]
We have now synced 86.7k crates (2.3m versions) according to https://sourcegraph.com/search?q=context:global+repo:%5Ecrates/+count:all&patternType=lucky

However, we disabled continuous syncing due to our excessive usage of the crates.io API. I just merged https://github.com/sourcegraph/sourcegraph/pull/38074 which fixes the API issues. I'm waiting for that PR to get deployed in gitserver and after that we can re-enable continuous syncing.

**Comment by olafurpg** [comment]
I've re-enabled continuous syncing with crates.io, which was the last remaining checkbox to complete this issue

**Comment by olafurpg** [comment]
Reopening this issue since there appears to be a non-trivial lag for crate versions to sync with Sourcegraph. All crates should sync every 12h but I suspect we're syncing stale commits from crates.io-index due to https://github.com/sourcegraph/sourcegraph/issues/37690 

I'm also open to adding a special-case when visiting the `/compare` page so that we fetch the versions directly from crates.io, that would allow us to guarantee that we always all version available, including ones that got published within the last <12h

**Comment by olafurpg** [comment]
Small update: I'm working on a branch that makes the `/compare` page sync the requested versions so that it's guaranteed to load as long as crates.io has the version available. I've validated that this approach is possible, just need to clean up the implementation (get a green CI, add tests).

**Comment by olafurpg** [comment]
I've opened a PR https://github.com/sourcegraph/sourcegraph/issues/37921 that implements the strategy above, which is to fallback to an on-demand sync when the tries to load a version that we haven't synced yet. The PR description includes a video explaining the change in more details.

**Comment by olafurpg** [comment]
Updated the checklist in the description to also include https://github.com/sourcegraph/sourcegraph/issues/37690 since I observed that we're syncing stale commits from rust-lang/crates.io. Once the PR #38811 is merged then we make it possible to sync new versions for *existing* crates on-demand when the user visits a URL, while #37690 is needed to pick up the creation of newly published crates (the first released version).

**Comment by olafurpg** [comment]
We just merged #37690 that should make all crate versions available on Sourcegraph with at most a 12-14hr delay. Previously, we didn't run a `git fetch` on rust-lang/crates.io-index so we were syncing with stale commits. 

The PR https://github.com/sourcegraph/sourcegraph/pull/38811 additionally makes it so that Sourcegraph syncs missing versions on-demand when a URL like https://sourcegraph.com/crates/wasm-encoder@v0.14.0 is visited (even if that version was release 10 minutes ago). This does not work for newly published crates, it only works for crates that have at least one version that's already tracked on Sourcegraph. It's possible to auto-sync new crates as well, but that will require a separate PR to implement.

I estimate both fixes will be live around middle of next week.

**Comment by olafurpg** [comment]
Reopening this issue to resolve the caveat that we don't lazy-sync new crates that have no previous versions. I have a WIP branch that fixes this issue so that Sourcegraph should never return a 404 when a crate is available on crates.io

**Comment by olafurpg** [comment]
Draft PR to receive notifications for updates https://github.com/sourcegraph/sourcegraph/pull/39179

**Comment by olafurpg** [comment]
The PR #39179 has been deployed and I just confirmed that I was able to load a crate that published its first version 23 minutes ago https://sourcegraph.com/crates/cliparser@v0.1.0

You can try it yourself by visiting the URL of a crate from the "New crates" list on https://crates.io
